### PR TITLE
Abstract search list from 'Add Token' for general use, and add new token search list story

### PR DESCRIPTION
**Pull request by danjm**
This PR does the following:

- refactors the `renderSearchToken` method of `add-token.component` - and its subcomponents `TokenSearch` and `TokenList` - to a general purpose search list, now found in `ui/app/components/ui/searchable-item-list/`
    - The new `SearchableItemList` component is not connected to redux state; all of its data is received from its parent.
    - This refactor includes the removal of the distinction between `tokens`, `selectedTokens` and `results` in the `TokenSearch` and `TokenList` logic. Instead the items in the list to search passed to `SearchableItemList` can have `selected` and `disabled` flags. 
    - All the components within `ui/app/components/ui/searchable-item-list/` follow our latest functional component style
- Refactors `add-token.component` to use the new `SearchableItemList`. This required `AddToken` to now handle all token specific log that was previously with `TokenSearch` and `TokenList`. This removes `selectedTokens` and `searchResults` from `AddToken`'s internal state.
- Adds a storybook story for the `AddToken` list to `ui/app/components/ui/searchable-item-list/searchable-item-list.stories.js`
- Adds a storybook story for a new searchable token list design to `ui/app/components/ui/searchable-item-list/searchable-item-list.stories.js` and adds scss for this story in `ui/app/pages/token/index.scss`

Here is a gif of the add token search list:
add-token-search

Here is a gif of the new searchable token list:

new-token-search-list

### CODE REVIEW

**Review by Gudahtt — CHANGES_REQUESTED** [review]

File: test/e2e/metamask-ui.spec.js [hunk 1]
```diff
@@ -750,7 +750,7 @@ describe('MetaMask', function () {
     })
 
     it('renders the correct ETH balance', async function () {
-      const balance = await driver.findElement(By.css('[data-testid="eth-overview__primary-currency"]'))
```
[the remark this answers is not in the document]
Comment (danjm) [reply test/e2e/metamask-ui.spec.js:753]: oh, darn, it looks like I made some sort of mistake while rebasing... this change was not intended, nor was the change you commented on here https://github.com/MetaMask/metamask-extension/pull/8877#discussion_r454396234

File: ui/app/components/ui/searchable-item-list/item-list/item-list.component.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import classnames from 'classnames'
+
+export default function SearchableItemList ({
+  results = [],
```
[the remark this answers is not in the document]
Comment (danjm) [reply ui/app/components/ui/searchable-item-list/item-list/item-list.component.js:6]: I personally prefer use of default parameters: fewer lines of code, and having the variable declaration and default setting in the same place results in less moving around when reading the code (with these functional components the `defaultProps` object can be many lines away from the function signature)

[the remark this answers is not in the document]
Comment (danjm) [reply ui/app/components/ui/searchable-item-list/item-list/item-list.component.js:6]: Does the fact that "Most other components use defaultProps to set default props instead of this syntax" mean that there is already an established preference on the team? If so, I'll follow that, but if not I'll campaign for using default parameters.

File: test/e2e/metamask-ui.spec.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (danjm) [reply test/e2e/metamask-ui.spec.js:753]: Fixed in the rebased b56ede5cb

File: app/images/storybook-icons/0x.svg [hunk 3]
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100"><title>Asset 1</title><g id="Layer_2" data-name="Layer 2"><g id="Layer_1-2" data-name="Layer 1"><path d="M24.9,93.22a49.86,49.86,0,0,0,52-1.09A50.18,50.18,0,0,0,88.32,82.07c-2.72-3.79-5.62-7.7-8.68-11.68-.82-1.07-1.67-2.15-2.52-3.24A32.27,32.27,0,0,1,63.55,79.09L55,70.73Z"/><path d="M6.63,25.16A49.87,49.87,0,0,0,7.88,76.89a50.54,50.54,0,0,0,10,11.41c3.79-2.72,7.71-5.62,11.69-8.68l3.24-2.51A32.22,32.22,0,0,1,20.91,63.54l8.42-8.62Z"/><path d="M75.12,6.79a49.87,49.87,0,0,0-52,1.09,50.54,50.54,0,0,0-11.41,10c2.72,3.79,5.62,7.71,8.68,11.69.82,1.07,1.67,2.16,2.51,3.24A32.31,32.31,0,0,1,36.46,20.9h0l8.05,7.85Z"/><path d="M93.48,74.63a49.82,49.82,0,0,0-1.35-51.5A50.18,50.18,0,0,0,82.07,11.68c-3.79,2.72-7.7,5.62-11.68,8.68-1.07.82-2.15,1.67-3.24,2.52a32.27,32.27,0,0,1,12,13.6l0,.12L71.05,45Z"/></g></g></svg>
```
[the remark this answers is not in the document]
Comment (danjm) [reply app/images/storybook-icons/0x.svg]: Moved them to a directory that will only be added to the storybook build in 97a31fe72

File: ui/app/components/ui/searchable-item-list/list-item-search/index.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+import TokenSearch from './list-item-search.component'
+
+export default TokenSearch
```
[the remark this answers is not in the document]
Comment (danjm) [reply ui/app/components/ui/searchable-item-list/list-item-search/index.js]: fixed in 6d40d7054

File: ui/app/components/ui/searchable-item-list/item-list/item-list.component.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply ui/app/components/ui/searchable-item-list/item-list/item-list.component.js:6]: I prefer default parameters for the same reasons. I thought there were reasons to prefer `defaultProps` regardless, but I can't recall any, and the rest of the team seems OK with using default parameters as well. Plus the React team is planning on deprecating `defaultProps` for functional components, so I guess we should consider the use of default parameters to be convention from this point forward.

**Comment by Gudahtt** [comment]
@danjm I can review this again if you rebase, now that we're past most of the v8 bugs.

**Comment by danjm** [comment]
@Gudahtt This has been rebased.

**Comment by Gudahtt** [comment]
The build is still failing

File: ui/app/components/ui/searchable-item-list/item-list/item-list.component.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            results.slice(0, maxListItems)
+              .map((result, i) => {
+                const {
+                  backgroundImageUrl,
+                  selected,
+                  disabled,
+                  primaryLabel,
+                  secondaryLabel,
+                  rightPrimaryLabel,
+                  rightSecondaryLabel,
```
Comment (Gudahtt) [inline ui/app/components/ui/searchable-item-list/item-list/item-list.component.js:33]: Are these used anywhere currently (aside from in Storybook)? Or are they solely for future use?

File: ui/app/pages/token/index.scss [hunk 6]
```diff
@@ -0,0 +1,148 @@
+.token {
```
Comment (Gudahtt) [inline ui/app/pages/token/index.scss:1]: Was this meant to go on the  `add-token` page? It's just a lone `.scss` file at the moment.

Comment (danjm) [reply ui/app/pages/token/index.scss:1]: This is a new directory for an upcoming feature. Currently this css is used in `/home/danjm/kyokan/metaswaps/ui/app/components/ui/searchable-item-list/searchable-item-list.stories.js`

File: ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+      position="start"
+      style={{ marginRight: '12px' }}
+    >
+      <img src="images/search.svg" />
+    </InputAdornment>
+  )
+
+  return (
+    <TextField
+      id="search-list-items"
```
[the remark this answers is not in the document]
Comment (danjm) [reply ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js]: Addressed in 5a56e1194

File: ui/app/pages/add-token/add-token.component.js [hunk 8]
```diff
@@ -31,33 +33,29 @@ class AddToken extends Component {
...
 
   componentDidMount () {
     this.tokenInfoGetter = tokenInfoGetter()
+    const { tokensToSearch } = this.state
     const { pendingTokens = {} } = this.props
-    const pendingTokenKeys = Object.keys(pendingTokens)
-
-    if (pendingTokenKeys.length > 0) {
-      let selectedTokens = {}
+    const pendingTokensCopy = { ...pendingTokens }
```
[the remark this answers is not in the document]
Comment (danjm) [reply ui/app/pages/add-token/add-token.component.js]: Addressed in 29756bb66

File: ui/app/pages/add-token/add-token.component.js [hunk 9]
```diff
@@ -31,33 +33,29 @@ class AddToken extends Component {
...
-      pendingTokenKeys.forEach((tokenAddress) => {
-        const token = pendingTokens[tokenAddress]
+      Object.entries(pendingTokensCopy).forEach(([tokenAddress, token]) => {
         const { isCustom } = token
-
         if (isCustom) {
           customToken = { ...token }
-        } else {
-          selectedTokens = { ...selectedTokens, [tokenAddress]: { ...token } }
+        } else if (!checkExistingAddresses(tokenAddress, newTokensToSearch)) {
```
[the remark this answers is not in the document]
Comment (danjm) [reply ui/app/pages/add-token/add-token.component.js]: You are right that it is not possible to hit that state. I removed all unnecessary changes to the `componentDidMount` in 29756bb66.

I then moved the setting of custom token props as you suggested in bde75d861.

File: ui/app/pages/add-token/add-token.component.js [hunk 10]
```diff
@@ -100,8 +92,8 @@ class AddToken extends Component {
   }
 
   hasSelected () {
-    const { customAddress = '', selectedTokens = {} } = this.state
-    return customAddress || Object.keys(selectedTokens).length > 0
+    const { customAddress = '', tokensToSearch } = this.state
+    return customAddress || tokensToSearch.find((tokenToSearch) => tokenToSearch.selected)
```
[the remark this answers is not in the document]
Comment (danjm) [reply ui/app/pages/add-token/add-token.component.js]: addressed in 1bde8e648

File: ui/app/css/design-system/colors.scss [hunk 11]
```diff
@@ -61,6 +61,8 @@ $Yellow-100: #fefcde;
 $Yellow-500: #ffd33d;
 
 $Black-100: #24292e;
+$Black-500: #6a737d;
```
[the remark this answers is not in the document]
Comment (danjm) [reply ui/app/css/design-system/colors.scss]: addressed in f472ec965

File: ui/app/components/ui/searchable-item-list/item-list/item-list.component.js [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+          <div className="searchable-item-list__title">
+            { listTitle }
+          </div>
+        )}
+        <div className="searchable-item-list__list-container">
+          {
+            results.slice(0, maxListItems)
+              .map((result, i) => {
+                const {
+                  backgroundImageUrl,
```
[the remark this answers is not in the document]
Comment (danjm) [reply ui/app/components/ui/searchable-item-list/item-list/item-list.component.js]: addressed in 9008b8dc5

File: ui/app/components/ui/searchable-item-list/item-list/item-list.component.js [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+                  selected,
+                  disabled,
+                  primaryLabel,
+                  secondaryLabel,
+                  rightPrimaryLabel,
+                  rightSecondaryLabel,
+                } = result
```
[the remark this answers is not in the document]
Comment (danjm) [reply ui/app/components/ui/searchable-item-list/item-list/item-list.component.js:34]: addressed in 059aea5a0

File: ui/app/components/ui/searchable-item-list/item-list/item-list.component.js (hunk 5 above)
Comment (danjm) [reply ui/app/components/ui/searchable-item-list/item-list/item-list.component.js:33]: For future use

**Review by danjm — COMMENTED** [review]
@Gudahtt I have addressed all your comments

File: ui/app/pages/add-token/add-token.component.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply ui/app/pages/add-token/add-token.component.js]: It looks like this remains unaddressed - `customAddress` is still returned directly, rather than cast as a boolean.

File: ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+const generateFuseKeys = (keys) => {
+  const weight = 1 / keys.length
+  return keys.map((key) => ({ name: key, weight }))
+}
+
+function usePrevious (value) {
+  const ref = useRef()
+  useEffect(() => {
+    ref.current = value
+  })
```
Comment (Gudahtt) [inline ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js:16]: This is missing the `value` dependency:

```suggestion
  }, [value])
```

File: ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+        location: 0,
+        distance: 100,
+        maxPatternLength: 32,
+        minMatchCharLength: 1,
+        keys: fuseSearchKeys || generateFuseKeys(searchByKeys),
+      })
+    }
+  }, [ searchByKeys, fuseSearchKeys, listToSearch ])
+
+  const previous = usePrevious({ listToSearch }) || []
```
Comment (Gudahtt) [inline ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js:58]: I don't think an _array_ is a very good default for this data structure, which is otherwise always an object.
```suggestion
  const previous = usePrevious({ listToSearch }) || {}
```

Though... why is this an object :thinking:

**Review by Gudahtt — CHANGES_REQUESTED** [review]

File: ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+  useEffect(() => {
+    if (didMountRef.current && searchQuery && previous.listToSearch !== listToSearch) {
+      fuse.setCollection(listToSearch)
+      const fuseSearchResult = fuse.search(searchQuery)
+      onSearch({ searchQuery, results: fuseSearchResult })
+    }
+  }, [ listToSearch, searchQuery, onSearch, previous.listToSearch ])
+
+
+  const renderAdornment = () => (
```
[the remark this answers is not in the document]
Comment (danjm) [reply ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js]: Done in https://github.com/MetaMask/metamask-extension/pull/8877/commits/349280d1070a3be1c6ef2d8994514dc1f782c941

File: ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+    })
+  }
+
+  const didMountRef = useRef(false)
+  useEffect(() => {
+    if (!didMountRef.current) {
+      didMountRef.current = true
+      fuse = new Fuse(listToSearch, {
+        shouldSort: true,
+        threshold: 0.1,
```
[the remark this answers is not in the document]
Comment (danjm) [reply ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js]: oops, that was just me experimenting at some point, nice catch, fixed in https://github.com/MetaMask/metamask-extension/pull/8877/commits/349280d1070a3be1c6ef2d8994514dc1f782c941

File: ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+  return ref.current
+}
+
+let fuse
+
+export default function ListItemSearch ({
+  onSearch = null,
+  error = '',
+  listToSearch = [],
+  searchByKeys = [],
```
[the remark this answers is not in the document]
Comment (danjm) [reply ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js]: removed `searchByKeys` in https://github.com/MetaMask/metamask-extension/pull/8877/commits/349280d1070a3be1c6ef2d8994514dc1f782c941

File: ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+    const fuseSearchResult = fuse.search(newSearchQuery)
+    onSearch({
+      searchQuery: newSearchQuery,
+      results: defaultToAll && newSearchQuery === '' ? listToSearch : fuseSearchResult,
+    })
+  }
+
+  const didMountRef = useRef(false)
+  useEffect(() => {
+    if (!didMountRef.current) {
```
[the remark this answers is not in the document]
Comment (danjm) [reply ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js]: The code was refactored slightly in https://github.com/MetaMask/metamask-extension/pull/8877/commits/349280d1070a3be1c6ef2d8994514dc1f782c941, but what you said about `fuseSearchKeys` holds true: the fuse instance won't update when it changes.

I think that is okay though? The old components didn't support updating the fuse instance on a change of keys. I can't think of a use case where the key configuration would need to be updated after the fuse instance is created.

File: ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+
+function usePrevious (value) {
+  const ref = useRef()
+  useEffect(() => {
+    ref.current = value
+  })
+  return ref.current
+}
+
+let fuse
```
[the remark this answers is not in the document]
Comment (danjm) [reply ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js]: Fixed in https://github.com/MetaMask/metamask-extension/pull/8877/commits/349280d1070a3be1c6ef2d8994514dc1f782c941

File: ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+
+export default function ListItemSearch ({
+  onSearch = null,
+  error = '',
+  listToSearch = [],
+  searchByKeys = [],
+  fuseSearchKeys = null,
+  searchPlaceholderText = '',
```
[the remark this answers is not in the document]
Comment (danjm) [reply ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js]: I have removed the default for `fuseSearchKeys` and made that prop required. Why is a default for `searchPlaceholderText` not useful? What if someone wants a search list without placeholder text in the search bar?

File: ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js (hunk 14 above)
Comment (danjm) [reply ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js]: https://github.com/MetaMask/metamask-extension/pull/8877/commits/349280d1070a3be1c6ef2d8994514dc1f782c941

File: ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js (hunk 15 above)
Comment (danjm) [reply ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js]: fixed in https://github.com/MetaMask/metamask-extension/pull/8877/commits/349280d1070a3be1c6ef2d8994514dc1f782c941

**Comment by danjm** [comment]
@Gudahtt I have either addressed all your latest comments in the latest commit, or left a couple of replies to things I think didn't need to be fixed.

Still works as expected:

searchlistdemo

File: ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js]: The old component didn't take these as props though. Having props that are ignored after mount seems a bit hazardous.

File: ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply ui/app/components/ui/searchable-item-list/list-item-search/list-item-search.component.js]: Then they can omit this prop. The default being applied here doesn't help that case - `placeholder` is not a required parameter of `TextField`.

File: ui/app/components/ui/searchable-item-list/searchable-item-list.js [hunk 22]
```diff
@@ -0,0 +1,60 @@
+import React, { useState } from 'react'
+import PropTypes from 'prop-types'
+import ItemList from './item-list'
+import ListItemSearch from './list-item-search'
+
+export default function SearchableItemList ({
+  itemSelectorError = '',
```
Comment (Gudahtt) [inline ui/app/components/ui/searchable-item-list/searchable-item-list.js:7]: Most of these default values are pointless (i.e. things work identically without them).

I'd strongly prefer that we reduce the complexity where possible by getting rid of these branches that don't have any effect.

Comment (Gudahtt) [reply ui/app/components/ui/searchable-item-list/searchable-item-list.js:7]: `ItemList` has a good few useless defaults as well

File: ui/app/pages/add-token/add-token.component.js [hunk 23]
```diff
@@ -274,31 +261,22 @@ class AddToken extends Component {
...
         <Tab name={this.context.t('search')}>
-          { this.renderSearchToken() }
+          <SearchableItemList
+            itemsToSearch={transformTokensForSearchList(tokensToSearch)}
+            itemSelectorError={tokenSelectorError}
+            onClickItem={(token) => this.handleToggleToken(token)}
+            Placeholder={TokenListPlaceholder}
+            className="add-token__search-token"
+            searchPlaceholderText={this.context.t('searchTokens')}
+            fuseSearchKeys={[{ name: 'name', weight: 0.499 }, { name: 'symbol', weight: 0.499 }, { name: 'address', weight: 0.002 }]}
```
Comment (Gudahtt) [inline ui/app/pages/add-token/add-token.component.js:277]: Were these changes intentional? The old keys used were as follows:

```suggestion
            fuseSearchKeys={[{ name: 'name', weight: 0.5 }, { name: 'symbol', weight: 0.5 }]}
```

File: ui/app/pages/token/index.scss [hunk 24]
```diff
@@ -1,3 +1,152 @@
...
+  &__search-token {
+    margin: 24px;
+    border: 1px solid #fff;
+    box-sizing: border-box;
+    box-shadow: 0 0 14px rgba(0, 0, 0, 0.18);
+    border-radius: 6px;
+    min-height: 297px;
+    width: 100%;
+
+    .searchable-item-list {
```
Comment (Gudahtt) [inline ui/app/pages/token/index.scss:15]: I don't think this is a good idea. The class `.searchable-item-list` is an internal implementation detail of the `SearchableItemList` component. Using the class directly outside of that component dramatically complicates refactoring.

Additional styles should be passed in via an optional custom classname.

File: ui/app/pages/token/index.scss (hunk 6 above)
Comment (Gudahtt) [reply ui/app/pages/token/index.scss:5]: It seems unnecessary to include additional features and styles in what is already quite a large refactoring PR.

File: ui/app/pages/token/index.scss [hunk 25]
```diff
@@ -1,3 +1,152 @@
...
+
+      &__list-container {
+        display: flex;
+        flex-direction: column;
+        max-height: 330px;
+        overflow-y: scroll;
+        scrollbar-width: none;
+
+        &::-webkit-scrollbar {
+          display: none;
```
Comment (Gudahtt) [inline ui/app/pages/token/index.scss:40]: Given the recent confusion we've seen around hidden scrollbars, I don't think this is something we should do anymore in general (e.g. https://consensys.slack.com/archives/G4V2HTG0Y/p1595014237319800)

File: ui/app/pages/add-token/index.scss [hunk 26]
```diff
@@ -17,10 +17,74 @@
 
   &__search-token {
     padding: 16px;
-  }
 
-  &__token-list {
-    margin-top: 16px;
+    .searchable-item-list {
```
Comment (Gudahtt) [inline ui/app/pages/add-token/index.scss:21]: The same comment applies here - we shouldn't be referencing internal styles

File: ui/app/pages/add-token/add-token.component.js (hunk 23 above)
Comment (danjm) [reply ui/app/pages/add-token/add-token.component.js:277]: These changes were intentional. They make a small UX improvement that allows the user to copy and paste a token address into the list and find that token. However, now that you ask, I realize this concern should be separate from this PR, so I will remove it

**Comment by danjm** [comment]
Closing for now, will return to this once we are ready to update the styles of the add token component
